### ISSUE TITLE
Inconsistent Response Models

### ISSUE THREAD

**Issue by tstorek** [orchestracities/ngsi-timeseries-api#476]
related to #469

**Describe the bug**
Working with QuantumLeap I discovered that the response models seem to be inconsistent making it tedious to parse them.
Why is sometimes 'id' and 'type' used and the 'entityId' and 'entityType' in other places? This requires a lot of unnecessary extra models and translation for client creation. 

**To Reproduce**
If you call the /v2/entities you will receive the entity data as a last of the following:
```
[
      {
        "index": [
          "2021-04-21T11:00:00.463000+00:00"
        ],
        "id": "Kitchen",
        "type": "Room"
      }
      {
        "index": [
          "2021-03-27T15:49:13.874000+00:00"
        ],
        "id": "MyId",
        "type": "MyType"
      }
]
```
But when querying e.g. v2/entities/{entityId} you receive this one:
```
      {
        "index": [
          "2021-04-21T10:53:25.095000+00:00",
          "2021-04-21T11:00:00.463000+00:00"
        ],
        "entityId": "Kitchen",
        "attributes": [
          {
            "values": [
              20.0,
              20.0
            ],
            "attrName": "temperature"
          }
        ],
      }
```

**Expected behavior**
I expect that entityId or id is used in both models. However, I think that entityId would be the better choice.

**Environment (please complete the following information):**
 - OS: Ubuntu 20.04 LTS
 - Python. 3.8 with requests
 - Version 0.8.x

**Comment by tstorek** [comment]
Hint: This is not only for this example. Also for the /v2/entity endpoints the Response includes the key "attributes", how is this different from "attrs" in the attributes endpoint. I do  not Understand, why the models are not reused.

**Comment by c0c0n3** [comment]
@tstorek thank you so much for pointing this out. Yes there are a bunch of things that should change in the query data structures. I'd love to have a more uniform interface too, but we can't do this right now since that's an API change that would break backward compatibility, probably something we can tackle when we release `v 1.0`. Also, there's more serious issues w/ queries than annoying data structures:

- https://github.com/orchestracities/ngsi-timeseries-api/wiki/The-Mother-of-all-Queries

Again, we can't tackle any of those issues right now, but queries are definitely something we'll have to improve going forward.

>  I do not Understand, why the models are not reused

The code to generate Python dictionaries out of SQL result sets is a bit tangled and needs refactoring, but at the moment we don't have enough dev cycles to spend on this. Hint: give us a hand! :-)

**Comment by tstorek** [comment]
@c0c0n3 Thanks, for the hint. I am familiar with flask I actually cannot figure out how quantumleap actually builds its endpoints. I find the structure really difficult to grasp. Nevertheless, I am happy that this piece of software exists.

However, it has been already proposed in the issues #399. Moving to fastapi would really help to break up thise complicated structure and btw it will solve the issue of inconsistent models and documentation by autogenerating all these things. Performancewise it natively supports async function. This would be great advantage over Flask. 

In a nutshell, for well documented opensource project I would not like to miss fastapi anymore. Since, it is far easier to grasp as foreign developer.

**Comment by c0c0n3** [comment]
Hi @tstorek :-)

> I find the structure really difficult to grasp.

Tell me about it. I too have had my share of grief wading through the code :-) But yah, seriously, we know the code ain't tasty, well unless you like spagheti, and would need a rewrite to make it modular. Problem is, we're always hard-pressed to deliver new functionality and there isn't really time to refactor. So we whip something together to do the job even though we know it isn't ideal and hope we'll find the time to refactor it later. But like I said, that time never comes, so we're left with brittle code that's difficult to understand.

> Moving to fastapi would really help to break up these complicated structure

Maybe, but I don't see this as being a huge win in terms of modularity. It could make the binding of HTTP calls to Python methods/functions cleaner but at the end of the day, the real modularity issues are within the QL code, i.e. the methods/functions bound to the HTTP calls and, in turn, the method/functions they call.

> and btw it will solve the issue of inconsistent models and documentation by autogenerating all these things.

Sure, but we'd have to rewrite most of the API methods which like I said we have no time for.

> Performance-wise it natively supports async function. This would be great advantage over Flask.

Well, yes but just plonking down FastAPI wouldn't really make performance any better in QL. In most calls, framework overhead (GUnicorn + Flask) gets dwarfed by actual QL + DB processing time. You can try running the benchmarks yourself and do data analysis as explained here:

- https://github.com/orchestracities/ngsi-timeseries-api/wiki/Gauging-Performance

So just switching to FastAPI without rewriting QL internals to take advantage of asynchrony wouldn't really help much with performance. That's not to say moving to an async architecture wouldn't make performance better: most (if not all) QL tasks are IO bound so if we moved away from the current synchronous architecture we could gain quite a bit of extra performance. But like I said, we'd also need to rewrite quite a bit of the QL internals...

**Comment by tstorek** [comment]
@c0c0n3 well, it is true that the spagehtti behind the scene will probably last a little longer, but at least the docs would be clean and one defenetly knows whre to look for models and docs. Since, I also noticed that the swagger docs apparently are not connected with the application itself. Hence, the gui-interface because somehow useless but the models.

Is there a way for a personal exchange on this?

**Comment by c0c0n3** [comment]
@tstorek 

> but at least the docs would be clean and one defenetly knows whre to look for models and docs

yep, you do have a good point there :-)

> Is there a way for a personal exchange on this?

you're welcome to drop me a line at  `x.y@martel-innovate.com` where x=andrea and y=falconi :-)

**Comment by pooja1pathak** [comment]
Hi @c0c0n3 , I would like to contribute on this issue. Please assign it to me.

PFA excel sheet for the dataModel of different APIs of QL, in which I have compared the response model and suggested the new model.
[dataModel.xlsx](https://github.com/orchestracities/ngsi-timeseries-api/files/6491912/dataModel.xlsx)
 
Please let me know if I can go in this direction and change the response of the APIs with new dataModel.

**Comment by chicco785** [comment]
thx @pooja1pathak i am going to have a look and come back to you

**Comment by c0c0n3** [comment]
@pooja1pathak @tstorek I've just sent you an email re: meeting to chat about all this and put together an implementation plan. @pooja1pathak I've used the email address on your GitHub profile, is that still current? Thanks guys!!

**Comment by pooja1pathak** [comment]
yes @c0c0n3 , I've received your email

**Comment by chicco785** [comment]
@pooja1pathak it seems that the only inconsistent message is `/v2/entities`, I saw in your proposal you want to comply the other messages with `/v2/entities`, but I believe that the other way around is less invasive (and faster to achieve).

so i would change `/v2/entities` to:

```json
[
      {
        "index": [
          "2021-04-21T11:00:00.463000+00:00"
        ],
        "entityId": "Kitchen",
        "entityType": "Room"
      }
      {
        "index": [
          "2021-03-27T15:49:13.874000+00:00"
        ],
        "entityId": "MyId",
        "entityType": "MyType"
      }
]
```

this fix can be done now regardless switch to pedantic. so @pooja1pathak feel free to propose a PR

**Comment by tstorek** [comment]
Hello,



I am sorry for the late response. I am glad to hear that something will happen.

Unfortunately, after next week I will be out of office for about 6 weeks.

Hence, we should either meet before or after.



Cheers,



Thomas



Von: c0c0n3 ***@***.***>
Gesendet: Dienstag, 18. Mai 2021 13:48
An: orchestracities/ngsi-timeseries-api ***@***.***>
Cc: Storek, Thomas ***@***.***>; Mention ***@***.***>
Betreff: Re: [orchestracities/ngsi-timeseries-api] Inconsistent Response Models (#476)



@pooja1pathak<https://github.com/pooja1pathak> @tstorek<https://github.com/tstorek> I've just sent you an email re: meeting to chat about all this and put together an implementation plan. @pooja1pathak<https://github.com/pooja1pathak> I've used the email address on your GitHub profile, is that still current? Thanks guys!!

**Comment by chicco785** [comment]
> yes @c0c0n3 , I've received your email.
> 
> We can have the discussion.
> I have checked the dataModel of the entity in **`Orion v2`**, where I found `id` and `type` is used in place of `entityId` and `entityType`. So, I have created another sheet which will match the data format of Orion.
> [dataModel_new.xlsx](https://github.com/orchestracities/ngsi-timeseries-api/files/6506872/dataModel_new.xlsx)

we agreed that when we are ready for 0.9 we will implement this change which is more invasive for clients.
